consensus/progpow/progpow.go

```diff
@@ -0,0 +1,539 @@
+		update:   make(chan struct{}),
+		hashrate: metrics.NewMeterForced(),
+	}
+	if config.PowMode == ModeShared {
+		progpow.shared = sharedProgpow
+	}
+	progpow.remote = startRemoteSealer(progpow, notify, noverify)
+	return progpow
+}
+
+// NewTester creates a small sized progpow PoW scheme useful only for testing
+// purposes.
+func NewTester(notify []string, noverify bool) *Progpow {
+	return New(Config{PowMode: ModeTest}, notify, noverify)
+}
+
+// NewFaker creates a progpow consensus engine with a fake PoW scheme that accepts
+// all blocks' seal as valid, though they still have to conform to the Quai
+// consensus rules.
+func NewFaker() *Progpow {
+	return &Progpow{
+		config: Config{
+			PowMode: ModeFake,
+			Log:     &log.Log,
+		},
+	}
+}
+
+// NewFakeFailer creates a progpow consensus engine with a fake PoW scheme that
+// accepts all blocks as valid apart from the single one specified, though they
+// still have to conform to the Quai consensus rules.
+func NewFakeFailer(fail uint64) *Progpow {
+	return &Progpow{
+		config: Config{
+			PowMode: ModeFake,
+			Log:     &log.Log,
+		},
+		fakeFail: fail,
+	}
+}
+
+// NewFakeDelayer creates a progpow consensus engine with a fake PoW scheme that
+// accepts all blocks as valid, but delays verifications by some time, though
+// they still have to conform to the Quai consensus rules.
+func NewFakeDelayer(delay time.Duration) *Progpow {
+	return &Progpow{
+		config: Config{
+			PowMode: ModeFake,
+			Log:     &log.Log,
+		},
+		fakeDelay: delay,
+	}
+}
+
+// NewFullFaker creates an progpow consensus engine with a full fake scheme that
+// accepts all blocks as valid, without checking any consensus rules whatsoever.
+func NewFullFaker() *Progpow {
+	return &Progpow{
+		config: Config{
+			PowMode: ModeFullFake,
+			Log:     &log.Log,
+		},
+	}
+}
+
+// NewShared creates a full sized progpow PoW shared between all requesters running
+// in the same process.
+func NewShared() *Progpow {
+	return &Progpow{shared: sharedProgpow}
+}
+
+// Close closes the exit channel to notify all backend threads exiting.
+func (progpow *Progpow) Close() error {
+	progpow.closeOnce.Do(func() {
+		// Short circuit if the exit channel is not allocated.
+		if progpow.remote == nil {
+			return
+		}
+		close(progpow.remote.requestExit)
+		<-progpow.remote.exitCh
+	})
+	return nil
+}
+
+// lru tracks caches or datasets by their last use time, keeping at most N of them.
+type lru struct {
+	what string
+	new  func(epoch uint64) interface{}
+	mu   sync.Mutex
+	// Items are kept in a LRU cache, but there is a special case:
+	// We always keep an item for (highest seen epoch) + 1 as the 'future item'.
+	cache      *simplelru.LRU
+	future     uint64
+	futureItem interface{}
+}
+
+// newlru create a new least-recently-used cache for either the verification caches
+// or the mining datasets.
+func newlru(what string, maxItems int, new func(epoch uint64) interface{}) *lru {
```

Comment:
Again also better to use standard library implementations that are tried and tested. Hopefully it's a simple import+swap, but if not we can talk about if its necessary before merge.

Reply:
Thats what I thought at first, and tried to make the change you suggest, but soon learned its not possible here.

This is actually a specialized LRU that returns two items. The cache for the current epoch as well as the cache for the next epoch. You'll notice all the "special logic" is built onto the `simple.Lru` which is a standard library. If I was to re-implement their dual cache logic (which we need, in order to handle epoch rotations in Ethash), I would end up building something very similar to what they built.